packages/filesystem/src/browser/filesystem-frontend-contribution.ts

```diff
@@ -82,23 +86,40 @@ export class FileSystemFrontendContribution implements FrontendApplicationContri
     registerCommands(commands: CommandRegistry): void {
         commands.registerCommand(FileSystemCommands.UPLOAD, new FileSelection.CommandHandler(this.selectionService, {
             multi: false,
-            isEnabled: selection => this.canUpload(selection),
-            isVisible: selection => this.canUpload(selection),
+            isEnabled: selection => this.menuItemEnabled(selection),
+            isVisible: selection => this.menuItemVisible(selection),
             execute: selection => this.upload(selection)
         }));
     }
 
-    protected canUpload({ fileStat }: FileSelection): boolean {
-        return !environment.electron.is() && fileStat.isDirectory;
+    //the top-left corner menu should always be enabled
+    protected menuItemEnabled(selection: FileSelection | undefined): boolean {
+        selection;
+        return !environment.electron.is();
+    }
+
+    //the right-click context menu should be invisible when the selected item is a regular file
+    protected menuItemVisible(selection: FileSelection | undefined): boolean {
+        //the only cases we can't upload are: electron environment, and fileSelection is a regular file
+        return !environment.electron.is() && (selection == undefined || selection.fileStat.isDirectory);
     }
 
     protected async upload(selection: FileSelection): Promise<void> {
         try {
-            const source = TreeWidgetSelection.getSource(this.selectionService.selection);
-            await this.uploadService.upload(selection.fileStat.uri);
-            if (ExpandableTreeNode.is(selection) && source) {
-                await source.model.expandNode(selection);
+            //use workspace root if there is no selection
+            if (selection == undefined) {
+                await this.uploadService.upload(this.workspaceService.tryGetRoots()[0].uri);
```

Comment:
FYI, this will upload to the first root of a multi-root workspace each time (if not selection is currently present)
Is this the behavior we want?

It might be worthwhile to prompt users to select the root they want to upload to in this case using a quick-pick menu.